sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala

```diff
@@ -135,19 +134,11 @@ abstract class AbstractGpuJoinIterator(
   private def nextCbFromGatherer(): Option[ColumnarBatch] = {
     withResource(new NvtxWithMetrics(gatherNvtxName, NvtxColor.DARK_GREEN, joinTime)) { _ =>
       val ret = gathererStore.map { gather =>
-        gather.checkpoint
-        gather.allowSpilling()
+        gather.checkpoint()
```

Comment:
nit, it seems this call should be part of the checkpoint/restore contract, so it shouldn't be outside IMHO. That said, it can be thought about as a follow on for sure.

Reply:
I see what you are saying, but in this particular case, the `gather.checkpoint()` might spill, and we need that to happen before we enter the retry loop.  I supposed I could explicitly call `gather.allowSpilling()` before, and have the `withRestoreOnRetry` call checkpoint at the top of each iteration, but I don't think that's typically what we want.  Usually you want to do a checkpoint before entering the retry loop, and only restore if something went wrong.  If we do refactor as @revans2 suggested, that might be an opportunity to revisit this.